### ISSUE TITLE
Feature request: UMC Help Pages

### ISSUE THREAD

**Issue by vgaming** [wesnoth/wesnoth#2376]
This feature request tries to address a specific problem of many wesnoth add-ons: lack of a good place to show documentation/manual/Help.

Work-arounds currently existing in the wild:
1. [set_menu_item] entries that look like "ABC Help", where ABC is add-on name (for example, A New Land map)
2. [message] at game start for each of the add-ons.
3. Chat messages (for example, Rav_XP_Mod)
4. Multi-click hidden pop-up (for example, Ageless). Via "select" event.
5. Objectives area

Let's analyse how good current solutions solve the problem.
1 won't work for observers (UPD: with a proper key, it will on 1.13. See comments below)
2 won't work after you started the game.
3 won't work after somebody types stuff in chat, also it can only be very short.
4 is a bit hard to access, and explain.
5 is for one add-on/map only.

So, what constraints I would like to have for Unified Help:
* It should be accessible to both observers and players, both active sides and waiting
* There should be a clean API for any add-on to add it's help entries on-the-fly
* The Help should be non-linear. We should allow at least structuring entries as a tree. (I think a tree-like structure is good enough.)
* There should be no constraints on the page-s internals. An arbitrary action or Lua function should be able to be called once the user chooses a particular entry. (Though the API docs for add-ons should explain that this might be OOS-unsafe.) The requirement to be able to call a Lua function is because some help contents would probably be best generated when you access it.

Thoughts?

**Comment by vgaming** [comment]
Possible temporary work-arounds: keep the structured Help data in a central place to which all (or many) add-on authors agree to. Then, each and every add-on should copy-paste the same code to make this structured Help be available in-game, or everyone should depend on a specific Add-On that does it.

This is a nasty move that might be used for 1.12. Or for 1.14, if there won't be any built-in solution (probably not as 1.14 should be released pretty soon). The main thing I'd like to discuss in the issue is the real Unified Help anyway, just mentioning alternatives for completeness sake.

**Comment by ProditorMagnus** [comment]
For 1.13 gfgtdf says workaround 1 works for observers.

**Comment by vgaming** [comment]
@ProditorMagnus I tried right now, it doesn't work for observers for this action:
```
	[set_menu_item]
		id=creepwars_observer_test
		description="observer test"
		[command]
			[lua]
				code = << print("it works!") >>
			[/lua]
		[/command]
	[/set_menu_item]
```
(But it does work for players.)

**Comment by ProditorMagnus** [comment]
Use unsynced menu item instead.

**Comment by vgaming** [comment]
@ProditorMagnus thank you, it does indeed work with `synced=no`. I guess that means that the value of a "Unified Help" reduces drastically (you can already achieve something almost similar with `synced=no`). Closing this issue then.

**Comment by CelticMinstrel** [comment]
I do want to allow addons to add their own help topics to the main help area, but that's on hold until we get around to converting that dialog to GUI2. Still, I think it could be a reason to reopen this issue; any objections?

**Comment by vgaming** [comment]
@CelticMinstrel I'm OK with re-opening considering the point of view you wrote.

**Comment by Wedge009** [comment]
Marking postponed pending Help conversion to gui2.

**Comment by jostephd** [comment]
Why should changing the GUI toolkit that implements the help screen block allowing add-ons to add content to that screen?

**Comment by Wedge009** [comment]
Just going by celmin's response on 2018-01-18. Feel free to change as you see fit.

**Comment by CelticMinstrel** [comment]
It doesn't need to block it. It's more that I personally don't feel the effort of extending a help system that's likely to be soon deprecated is worthwhile. If someone else wants to do it, then I don't really have an objection (though I'd much rather they put effort into the GUI2 help instead).

**Comment by jostephd** [comment]
My point is that the API to addon authors and the presentation layer are entirely separate. Even if we throw away the latter completely and rewrite it from the ground up, the former won't need to change at all. And if we can let addon authors add help entries _now_, I think that might well be worthwhile, if the alternative is that addon authors can't add help entries until GUI2 help is done, which might happen in 1.18, or might not.

**Comment by CelticMinstrel** [comment]
I think that's not actually true. The help currently uses a custom markup, but GUI2 help uses Pango markup (which would have some extensions to it specific to help, ideally). Apart from the markup though, you might be correct. (And the markup would be automatically migrated by the GUI2 help anyway.)

The current API is very extension-unfriendly though if I recall correctly - each topic specifies its subtopics, right? Reversing that (each topic specifies its parent topic) would make it easier for add-ons to extend it, I think.

**Comment by jostephd** [comment]
Pretty sure the extensibility is solvable. How hard can it be to implement a WML action in lua that the parent section id as a parameter and adds a child to that section?

**Comment by CelticMinstrel** [comment]
Well, I don't actually think that's the right approach... you shouldn't have to use Lua (or ActionWML) to add a help topic. Simply defining the topic should be sufficient for it to show up in help. There _could_ be some Lua API here, for example if you want a topic to be hidden until a certain condition, or if you want actual hidden topics that can only be shown from code. But for the basic case of simply adding a new topic? That should not require any coding.

**Comment by jostephd** [comment]
@CelticMinstrel What I said was "implement a WML action in Lua". _We_ would write Lua that implements a `function wesnoth.wml_actions.foo`. UMC maintainers won't need to code anything.

**Comment by CelticMinstrel** [comment]
No, what I'm saying is that UMC maintainers shouldn't need to use an ActionWML tag or a Lua function to add a topic to the help. They should be able to just add the `[topic]` tag (or whatever) at an appropriate location in their add-on and it should automatically be picked up. The `wml_actions` table doesn't come into play here.

Sure, we could _also_ add WML actions to dynamically add or modify topics. But that shouldn't be the core API for adding extra topics.

**Comment by jostephd** [comment]
Maybe wml_action wasn't the best way to phrase it, but yes, that's what I meant. UMC authors would add a WML tag that specifies the parent and it would just work. (because under the hood some code written by us would dynamically modify the parent, etc)

**Comment by CelticMinstrel** [comment]
Right, but that WML tag wouldn't have an underlying Lua implementation since it's not an ActionWML tag (or ConditionWML, `[effect]`, etc). It would just be read by the help parser.

**Comment by Wedge009** [comment]
#3382 also notes the wiki resource, although that's external to the game.

**Comment by babaissarkar** [comment]
Could be done now that the Help Browser is ported to GUI2. (#3653)

**Comment by CelticMinstrel** [comment]
As I mentioned earlier, in order to support this, we probably need to refactor how relationships between topics are defined:

> The current API is very extension-unfriendly though if I recall correctly - each topic specifies its subtopics, right? Reversing that (each topic specifies its parent topic) would make it easier for add-ons to extend it, I think.

**Comment by babaissarkar** [comment]
How does #1155 differ from this one? Looks like a duplicate.

**Comment by CelticMinstrel** [comment]
I think it's a duplicate.

By the way, it might be nice to also support custom help topic generators. For example, an RPG-style campaign with an inventory system could use this to generate topics for every item in the campaign.

**Comment by soliton-** [comment]
> The proof-of-concept should probably be to make the WoCopedia appear in the main help, but only when playing World Conquest.

Why is it required to only show when playing WC? IMO putting UMC Help pages in their own section but showing them always (unless they want to play preprocessor tricks) would be better.

**Comment by CelticMinstrel** [comment]
I guess it's not strictly required, but it does make sense IMO.
